rosetta/retriever/retriever.go

```diff
@@ -68,14 +73,144 @@ func (r *Retriever) Balances(network identifier.Network, block identifier.Block,
+
+		// This switch just skips all events except the ones we are explicitly interested in.
+		switch event.Type {
+		default:
+			continue
+		case flow.EventType(deposit):
+		case flow.EventType(withdrawal):
+		}
+
+		// Decode the event payload into a Cadence value and cast to Cadence event.
+		value, err := json.Decode(event.Payload)
+		if err != nil {
+			return nil, nil, fmt.Errorf("could not decode event: %w", err)
+		}
+		e, ok := value.(cadence.Event)
+		if !ok {
+			return nil, nil, fmt.Errorf("could not cast event: %w", err)
+		}
+
+		// Now we have access to the fields for the events; the first one is always
+		// the amount, the second one the address. The types coming from Cadence
+		// are not native Flow types, so we need to use primitive types first.
+		vAmount := e.Fields[0].ToGoValue()
+		uAmount, ok := vAmount.(uint64)
+		if !ok {
+			return nil, nil, fmt.Errorf("could not cast amount (%T)", vAmount)
+		}
+		vAddress := e.Fields[1].ToGoValue()
+		bAddress, ok := vAddress.([flow.AddressLength]byte)
+		if !ok {
+			return nil, nil, fmt.Errorf("could not cast address (%T)", vAddress)
+		}
+
+		// Then, we can convert the amount to a signed integer so we can invert
+		// it and the address to a native Flow address.
+		amount := int64(uAmount)
+		address := flow.Address(bAddress)
+
+		// For the witdrawal event, we invert the amount into a negative number.
+		if event.Type == flow.EventType(withdrawal) {
+			amount = -amount
+		}
+
+		// Now we have everything to assemble the respective operation.
+		op := object.Operation{
+			ID: identifier.Operation{
+				Index: uint(event.EventIndex),
+			},
+			RelatedIDs: nil,
+			Type:       "TRANSFER",
+			Status:     "COMPLETED",
+			AccountID: identifier.Account{
+				Address: address.String(),
+			},
+			Amount: object.Amount{
+				Value: strconv.FormatInt(amount, 10),
+				Currency: identifier.Currency{
+					Symbol:   dps.FlowSymbol,
+					Decimals: dps.FlowDecimals,
+				},
+			},
+		}
+
+		// We store all operations for a transaction together in a bucket.
+		buckets[event.TransactionID] = append(buckets[event.TransactionID], op)
+	}
+
+	// Finally, we batch all of the operations together into the transactions.
+	var transactions []*object.Transaction
+	for transactionID, operations := range buckets {
+		transaction := object.Transaction{
+			ID: identifier.Transaction{
+				Hash: transactionID.String(),
+			},
+			Operations: operations,
+		}
+		for _, op := range operations {
+			for index := range transaction.Operations {
+				if transaction.Operations[index].ID != op.ID {
+					transaction.Operations[index].RelatedIDs = append(transaction.Operations[index].RelatedIDs, op.ID)
+				}
+			}
+		}
+		transactions = append(transactions, &transaction)
+	}
+
+	// Now we just need to build the block.
+	block := object.Block{
+		ID: identifier.Block{
+			Index: header.Height,
+			Hash:  header.ID().String(),
+		},
+		ParentID: identifier.Block{
+			Index: header.Height - 1,
```

Comment:
Yes, I'm aware. However, we will never request the first block, because it was significantly before `mainnet-1` and before anything relevant was running on the Flow blockchain. It is worth thinking about how we can solve this in terms of cross-spork parents, though. Will create an issue.